Make the nzgd_map website run in a Docker container (#2)
ucgmsim/nzgd_map: readme.md

## Introduction

This repository contains the source code for the `nzgd_map` package. This is a web application
that enables access to analysis-ready data products derived from data hosted on the New Zealand Geotechnical 
Database (NZGD). This repository also contains files for building a Docker image that can be used to run the
`nzgd_map` package in a containerized environment.

## Installation

### Installing outside a Docker container

#### Installing an uneditable version

To install an uneditable version of `nzgd_map` outside a Docker container, simply enter the following 
command in a terminal:

> * `pip install "git+https://github.com/ucgmsim/nzgd_map#egg=nzgd_map"` 

Reviewer: ```suggestion
* `pip install "git+https://github.com/ucgmsim/nzgd_map"` 
```

* Push the image to Docker Hub (see the section below on logging in to Docker Hub)
    * `sudo docker push earthquakesuc/nzgd_map`

* Copy `nzgd_map.service` to the machine where you want to run the `nzgd_map` package. For example, on Mantle,
    * `cp nzgd_map.service /etc/systemd/system/`

* Reload the systemd unit files. For example, on Mantle,
    *  `sudo systemctl daemon-reload`

* Pull the latest Docker image from Docker Hub and run the `nzgd_map` container by starting the `nzgd_map` service.
    * If the service is not currently running, start it.
        * `sudo systemctl start nzgd_map`
    * If the service is already running, restart it.
        * `sudo systemctl restart nzgd_map`

* Check the status of the `nzgd_map` service to ensure it is running
    * `sudo systemctl status nzgd_map`

* Check the status of the `nzgd_map` container to ensure it is running
    * `sudo docker ps`

### Logging in to Docker Hub
Open a terminal and enter the following command:
`sudo docker login`

The terminal will show a message like the following:

    USING WEB BASED LOGIN
    To sign in with credentials on the command line, use 'docker login -u <username>'

>     Your one-time device confirmation code is: JGPX-PVKR

Reviewer: In general, don't make one-time device codes you actually generated publicly available as these can be used to generate more depending on the algorithm used.

>     Your one-time device confirmation code is: JGPX-PVKR

Reviewer: ```suggestion
    Your one-time device confirmation code is: XXXX-XXXX
```

not that it makes much difference in this case, but it is good practice.

**Explanation:**  
Runs this command before starting the service, which pulls the latest Docker image "earthquakesuc/nzgd_map".

**File line:**  
`ExecStart=/snap/bin/docker run --rm --name %n -p 7777:80 -v /mnt/mantle_data/nzgd_map:/usr/var/nzgd_map-instance earthquakesuc/nzgd_map`  

**Explanation:**  
Starts the Docker container using the pulled image with the following options:  
- **--rm:** Automatically removes the container when it exits.  
- **--name %n:** Names the container based on the service name.  
- **-p 7777:80:** Maps host port 7777 to container port 80.  
- **-v /mnt/mantle_data/nzgd_map:/usr/var/nzgd_map-instance:** Mounts a host directory into the container.

**File line:**  
`ExecStop=/snap/bin/docker stop %n`  

**Explanation:**  
Stops the Docker container gracefully when the service is stopped.

**File line:**  
`ExecStopPost=/snap/bin/docker rm %n`  

**Explanation:**  
Removes the container after stopping to clean up any remnants.

**File line:**

`[Install]`

**Explanation:**

Starts the `[Install]` section, which defines how the service integrates with the system boot process.

**File line:**  
`WantedBy=default.target`  

**Explanation:**  

> Specifies that the service should be started when the default system target (usually multi-user mode) is reached.

Reviewer: I'd rather see this as markdown comments inside a systemd unit block. That way you can just copy the contents and the comments stay with the service.

```ini
# Starts the [Unit] section, which contains metadata and defines relationships with other services.
[Unit]

# Provides a short description of the service as "NZGD Map Service".
Description=NZGD Map Service

# Ensures that this service starts only after the Docker daemon service has started.
After=snap.docker.dockerd.service

# Declares a dependency on the Docker daemon service, ensuring it is running before this service starts.
Requires=snap.docker.dockerd.service

# Starts the [Service] section, which controls how the service process executes and behaves.
[Service]

# Sets the start timeout to 0 seconds, meaning systemd will wait indefinitely for the service to start.
TimeoutStartSec=0

# Configures systemd to always restart the service if it stops for any reason.
Restart=always

# Runs this command before starting the service, which pulls the latest Docker image "earthquakesuc/nzgd_map".
ExecStartPre=/snap/bin/docker pull earthquakesuc/nzgd_map

# Starts the Docker container using the pulled image with the following options:
# - --rm: Automatically removes the container when it exits.
# - --name %n: Names the container based on the service name.
# - -p 7777:80: Maps host port 7777 to container port 80.
# - -v /mnt/mantle_data/nzgd_map:/usr/var/nzgd_map-instance: Mounts a host directory into the container.
ExecStart=/snap/bin/docker run --rm --name %n -p 7777:80 -v /mnt/mantle_data/nzgd_map:/usr/var/nzgd_map-instance earthquakesuc/nzgd_map

# Stops the Docker container gracefully when the service is stopped.
ExecStop=/snap/bin/docker stop %n
# Removes the container after stopping to clean up any remnants.
ExecStopPost=/snap/bin/docker rm %n

# Starts the [Install] section, which defines how the service integrates with the system boot process.
[Install]

# Specifies that the service should be started when the default system target (usually multi-user mode) is reached.
WantedBy=default.target
```

Reviewer: wrap that code in an `ini` code block.

Reviewer: Actually, I've somewhat changed my mind here. Just put the comments in the files themselves and then link to them in the README. That way they can't be out of sync.

Reviewer: See my large suggestion for this file.

**File line:**  
RUN pip install --no-cache-dir "git+https://github.com/ucgmsim/nzgd_map#egg=nzgd_map" uwsgi`  

**Explanation:**  
Installs the `nzgd_map` package directly from a GitHub repository without using the pip cache, ensuring the latest version 
for container deployment.

**File line:**  
`COPY nzgd.ini /nzgd.ini`  

**Explanation:**
Copies the `nzgd.ini` configuration file from the build context into the container’s root directory.

**File line:**
`COPY start.sh /start.sh`  

**Explanation:**  
Copies the `start.sh` script from the build context into the container’s root directory.

**File line:**  
`RUN chmod +x /start.sh`  

**Explanation:**  
Grants execute permission to the `start.sh` script so that it can be run as a command when the container starts.

**File line:**  
`COPY nginx.conf /etc/nginx/nginx.conf`  

**Explanation:**  
Copies the `nginx.conf` file from the build context into the container’s `/etc/nginx/` directory, replacing the default 
nginx configuration.

**File line:**  
`CMD /start.sh`  

**Explanation:**  
Sets the default command to run when the container starts; it executes the `start.sh` script, which starts

> nginx in the foreground.

Reviewer: Similarly here, update with comments:
```dockerfile
# Specifies that Alpine Linux be used as the base image for the Docker container.
# Ubuntu Linux can also be used, but the resulting images are approximately 10 times larger.
FROM python:3.12-alpine  

# Uses Alpine’s package manager `apk` to install the necessary packages (`git`, `nginx`, 
# `sqlite` libraries, and `uwsgi` for `Python3`) without caching, which helps keep the image size small.
RUN apk add --no-cache git nginx sqlite-libs uwsgi-python3  

# Creates the `/data` directory if it does not already exist.
RUN mkdir -p /data  

# Creates the `/usr/var/nzgd_map-instance` directory, which is used for storing 
# instance-specific data for the NZGD map service.
RUN mkdir -p /usr/var/nzgd_map-instance  

# Changes the ownership of the `/usr/var/nzgd_map-instance` directory (and its contents) 
# recursively to the user and group “nobody” for improved security.
RUN chown -R nobody:nobody /usr/var/nzgd_map-instance  

# Installs the `nzgd_map` package directly from a GitHub repository without using the pip cache, 
# ensuring the latest version for container deployment.
RUN pip install --no-cache-dir "git+https://github.com/ucgmsim/nzgd_map#egg=nzgd_map" uwsgi  

# Copies the `nzgd.ini` configuration file from the build context into the container’s root directory.
COPY nzgd.ini /nzgd.ini  

# Copies the `start.sh` script from the build context into the container’s root directory.
COPY start.sh /start.sh  

# Grants execute permission to the `start.sh` script so that it can be run as a command when the container starts.
RUN chmod +x /start.sh  

# Copies the `nginx.conf` file from the build context into the container’s `/etc/nginx/` directory, 
# replacing the default nginx configuration.
COPY nginx.conf /etc/nginx/nginx.conf  

# Sets the default command to run when the container starts; it executes the `start.sh` script, 
# which starts nginx in the foreground.
CMD /start.sh  
```

Reviewer: This should replace `docker/Dockerfile` and then this should be a link.

**File line:**  
`    include /etc/nginx/mime.types;  # <- Ensures proper MIME types!`  

**Explanation:**  
Includes the MIME types configuration file that maps file extensions to MIME types. This ensures that files are served 
with the correct Content-Type header.

**File line:**  
``    server {``  

**Explanation:**  
Starts a server block which defines the configuration for a virtual server handling HTTP requests.

**File line:**  
``           listen 80;``  

**Explanation:**  
Instructs the server to listen for incoming HTTP requests on port 80.

**File line:**  
``           server_name localhost;``  

**Explanation:**  
Sets the server name to "localhost". This server block will handle requests directed to the hostname "localhost".

**File line:**  
``           location /nzgd {``  

**Explanation:**  
Begins a location block that matches requests starting with the URI `/nzgd`. This block defines how requests to that 
path should be handled.

**File line:**  
``                    include uwsgi_params;``  

**Explanation:**  
Includes the uWSGI parameters file, which defines necessary parameters for communicating with a uWSGI server.

**File line:**  

> ``                    uwsgi_pass unix:/tmp/nzgd.sock;``  
> 
> **Explanation:**  
> Directs Nginx to forward matching requests to a uWSGI server through the Unix socket located at `/tmp/nzgd.sock`.
> 
> **File line:**  
> ``                    uwsgi_param SCRIPT_NAME /nzgd;``  
> 
> **Explanation:**  
> Sets the `SCRIPT_NAME` parameter to `/nzgd` so that the uWSGI application correctly interprets the base URI for requests.
> 
> **File line:**  
> ``           }``  
> 
> **Explanation:**  
> Closes the location block.
> 
> 
> **File line:**  
> ``    }``  
> 
> **Explanation:**  
> Closes the server block.
> 
> **File line:**  
> `}`  
> 
> **Explanation:**  
> Closes the HTTP configuration block.
> 
> ### start.sh
> 
> 
> **File line:**  
> `#!/usr/bin/env sh`  
> 
> **Explanation:**  
> Specifies the script interpreter by using the environment's `sh`. This is the shebang line that tells the system to 
> execute the script with the shell interpreter.
> 
> **File line:**  
> `echo "Starting nginx"`  
> 
> **Explanation:**  
> Prints the message "Starting nginx" to the console, which serves as an informational log to indicate that the script is
> beginning the process of starting Nginx.
> 
> **File line:**  
> `uwsgi --ini /nzgd.ini &`  
> 
> **Explanation:**  
> Starts the uWSGI server using the configuration file `/nzgd.ini` and runs it in the background 
> (due to the trailing `&`), allowing the script to continue executing subsequent commands.
> 
> **File line:**  
> `nginx -g "daemon off;"`  
> 
> **Explanation:**  
> Launches Nginx with the directive to run in the foreground (`daemon off;`). 
> Running Nginx in the foreground is particularly useful in container environments to prevent the container from exiting.

Reviewer: ```suggestion
The following files in the `docker` directory setup the NZGD map service in a container, and run it on startup with systemd. 

- [Service file to start NZGD map service](docker/nzgd_map.service)
- [Dockerfile defining NZGD map container](docker/Dockerfile)
- [uWSGI configuration for NZGD server](docker/nzgd.ini)
- [nginx config exposing server outside the container](docker/nginx.conf)
- [Entrypoint script that runs when container is executed](docker/start.sh)

```
